### PR TITLE
Remote value serialisations

### PR DESCRIPTION
**Pull request by sadym-chromium**
* `undefined`
* `null`
* `string`
* `number`
* `specialNumber`
* `bool`
* `bigint`
* `symbol`
* `function`
* `regExp`
* `date`
* `windowProxy`
* `error`
* `node`

### CODE REVIEW

File: bidiClient/test_bidi.py [hunk 1]
```diff
@@ -819,44 +660,247 @@ async def test_consoleError_logEntryWithMethodErrorRaised(websocket):
...
+# Testing serialisation.
+async def assertSerialisation(jsStrObject, expectedSerialisedObject, hasObjectId, websocket):
+    contextID = await get_open_context_id(websocket)
+
+    # Send command.
+    await send_JSON_command(websocket, {
+        "id": 9997,
+        "method": "PROTO.page.evaluate",
+        "params": {
+            "function": f"console.log({jsStrObject})",
```
Comment (foolip) [inline bidiClient/test_bidi.py:672]: Rather than using `console.log`, can't the expression be evaluated directly? This way there should be no need to wait for any event, only the response to the command.

File: bidiClient/test_bidi.py [hunk 2]
```diff
@@ -819,44 +660,247 @@ async def test_consoleError_logEntryWithMethodErrorRaised(websocket):
...
+async def test_serialisation_string(websocket):
+    await assertSerialisation(
+        "'someStr'",
+        {
+            "type":"string",
+            "value":"someStr"},
+        False, websocket)
+
+@pytest.mark.asyncio
+async def test_serialisation_number(websocket):
```
Comment (foolip) [inline bidiClient/test_bidi.py:716]: Can you also test a floating point number such as 0.5?

File: bidiClient/test_bidi.py [hunk 3]
```diff
@@ -819,44 +660,247 @@ async def test_consoleError_logEntryWithMethodErrorRaised(websocket):
...
+
+@pytest.mark.asyncio
+async def test_serialisation_null(websocket):
+    await assertSerialisation(
+        "null",
+        {"type":"null"},
+        False, websocket)
+
+@pytest.mark.asyncio
+async def test_serialisation_string(websocket):
```
Comment (foolip) [inline bidiClient/test_bidi.py:707]: Can you also test a few cases to exercise escaping? A string with null bytes in it would be great, as well as one with lone surrogates. It's possible Python websocket library we're using will modify the values or break in some way, in which case leaving TODOs would be fine.

File: bidiClient/test_bidi.py [hunk 4]
```diff
@@ -819,44 +660,247 @@ async def test_consoleError_logEntryWithMethodErrorRaised(websocket):
...
+            "value":False},
+        False, websocket)
+
+@pytest.mark.asyncio
+async def test_serialisation_bigint(websocket):
+    await assertSerialisation(
+        "BigInt('12345678901234567890')",
+        {
+            "type":"bigint",
+            "value":"12345678901234567890n"},
```
Comment (foolip) [inline bidiClient/test_bidi.py:775]: I think that the trailing "n" shouldn't be here. Trying `BigInt('12345678901234567890').toString()
` there's no trailing "n" and the spec is defined in terms of https://tc39.es/ecma262/#sec-tostring.

File: bidiClient/test_bidi.py [hunk 5]
```diff
@@ -819,44 +660,247 @@ async def test_consoleError_logEntryWithMethodErrorRaised(websocket):
...
+        False, websocket)
+
+@pytest.mark.asyncio
+async def test_serialisation_symbol(websocket):
+    await assertSerialisation(
+        "Symbol('foo')",
+        {"type":"symbol"},
```
Comment (foolip) [inline bidiClient/test_bidi.py:782]: This is correct, but it seems useful to also include the description. I've filed https://github.com/w3c/webdriver-bidi/issues/89.

File: bidiClient/test_bidi.py [hunk 6]
```diff
@@ -819,44 +660,247 @@ async def test_consoleError_logEntryWithMethodErrorRaised(websocket):
...
+        True, websocket)
+
+@pytest.mark.asyncio
+async def test_serialisation_function(websocket):
+    await assertSerialisation(
+        "function(){}",
+        {"type":"function"},
```
Comment (foolip) [inline bidiClient/test_bidi.py:789]: Does CDP include any other information about functions without a roundtrip? Maybe the `name` or `length` (number of required arguments) of the function?

File: bidiClient/test_bidi.py [hunk 7]
```diff
@@ -819,44 +660,247 @@ async def test_consoleError_logEntryWithMethodErrorRaised(websocket):
...
+    await assertSerialisation(
+        "new Error('Woops!')",
+        {"type":"error"},
+        True, websocket)
+
+@pytest.mark.asyncio
+async def test_serialisation_node(websocket):
+    await assertSerialisation(
+        "document.createElement('p')",
+        {"type":"node"},
```
Comment (foolip) [inline bidiClient/test_bidi.py:829]: Bits from `NodeProperties` are missing here. When you add that, testing every node type would also be great.

File: bidiClient/test_bidi.py (hunk 1 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:672]: Good idea! Unfortunately, the `PROTO.page.evaluate` doesn't use proper serialisation yet. Working on it.

Comment (foolip) [reply bidiClient/test_bidi.py:672]: Great! It will of course be important to test serialization of `console.log` too, but a handful of tests should suffice there, we need not test every combination of features.

File: bidiClient/test_bidi.py (hunk 2 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:716]: done in the next revision

File: bidiClient/test_bidi.py (hunk 3 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:707]: Added TODO.

File: bidiClient/test_bidi.py (hunk 4 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:775]: done in the next revision

File: bidiClient/test_bidi.py (hunk 5 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:782]: Added `PROTO.description` in the next revision

File: bidiClient/test_bidi.py (hunk 6 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:789]: It has a `description`, but not sure how useful will it be. CDP response:
```
{"type":"function","className":"Function","description":"function(){}","objectId":"4345989804952205736.1.1"}}
```

File: bidiClient/test_bidi.py (hunk 7 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:829]: Added `TODO`, but not clear how to implement it before we have a clear solution of https://github.com/w3c/webdriver-bidi/issues/86

File: bidiClient/test_bidi.py (hunk 1 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:672]: https://github.com/sadym-chromium/WebDriverBiDiServerExperiments/pull/40

File: bidiClient/test_bidi.py (hunk 2 above)
Comment (sadym-chromium) [reply bidiClient/test_bidi.py:716]: #40

File: bidiClient/test_bidi.py (hunk 6 above)
Comment (foolip) [reply bidiClient/test_bidi.py:789]: Yeah, that doesn't look useful. No name then at least.
